### PR TITLE
JS Components: Add spacing props to Text component

### PR DESCRIPTION
**Pull request by retrofox**
This PR implements a spacing feature for the `<Text />` component.

The new props supported allow define the box-model for the component (margin and padding, not border)
Prop | Desc
------|------
`m` |  margin
`mt` | margin-top
`mr` | margin-right
`mb` | margin-bottom
`ml` | margin-left
`mx` | margin left and right
`my` | margin top and bottom
`p` |  padding
`pt` | padding-top
`pr` | padding-right
`pb` | padding-bottom
`pl` | padding-left
`px` | padding left and right
`py` | padding top and bottom

Example:
```jsx
function MyBoxModel() {
  // margin-left: 24px
  // margin-right: 24px
  // padding-bottom: 48px
  return (
    <Text mx={ 3 } pb={ 6 }>Hello</Text>
  );
}
```

Fixes #

#### Changes proposed in this Pull Request:

* JS Components: Add spacing props to Text component

#### Testing instructions:

* Go to My Jetpack overview
* Confirm the title of Plans and Connection sections are correct
  * font-size: 36px
  * margin-bottom: 24px;
Screen Shot 2022-03-29 at 3 09 50 PM

* Using Storybook
* Select the Box Model story of the Text component
* Check the spacing properties

https://user-images.githubusercontent.com/77539/160678313-b3b82fde-fc7c-472d-86ed-380402163f8e.mov

### CODE REVIEW

File: projects/js-packages/components/components/text/index.jsx [hunk 1]
```diff
@@ -32,11 +32,41 @@ export const TIPOGRAPHY_SIZES = {
...
+export const SPACING_VALUES = [ 0, 1, 2, 3, 4, 5, 6, 7, 8 ];
+
 const Text = ( { variant, children, component, className, ...componentProps } ) => {
 	const Component = component || VARIANTS_MAPPING[ variant ] || 'span';
-	const componentClassName = classNames( styles[ variant ], className );
+
+	// Build Styles module CSS classnames.
+	const boxModelClasses = BOX_MODEL_VALUES.reduce( ( acc, value ) => {
+		if ( typeof componentProps[ value ] !== 'undefined' ) {
+			acc += styles[ `${ value }-${ componentProps[ value ] }` ] + ' ';
```
Comment (renatoagds) [inline projects/js-packages/components/components/text/index.jsx:59]: We need to get directly these props and not use them from `componentProps`, which is used for composition and normally applies directly to first-class tags `<h1>, <p>`, etc ... 

This way we'll probably receive some warnings

File: projects/js-packages/components/components/text/stories/index.jsx [hunk 2]
```diff
@@ -6,11 +6,56 @@ import React from 'react';
...
+		pl: {
+			control: { type: 'select', options: SPACING_VALUES },
+		},
+		px: {
+			control: { type: 'select', options: SPACING_VALUES },
+		},
+		py: {
+			control: { type: 'select', options: SPACING_VALUES },
+		},
+	},
```
Comment (renatoagds) [inline projects/js-packages/components/components/text/stories/index.jsx:58]: You could simplify this one exporting `BOX_MODEL_VALUES`:

```es6
argTypes: {
  ...BOX_MODEL_VALUES.reduce((obj, cur) => ({
    ...obj, [cur]: {
      control: { type: 'select', options: SPACING_VALUES },
    }
  }), {}),
}
```

File: projects/js-packages/components/components/text/style.module.scss [hunk 3]
```diff
@@ -70,14 +70,45 @@
...
-	margin-bottom: var( --spacing-base );
+$sides: (
+	"top": "t",
+	"right": "r",
+	"bottom": "b",
+	"left": "l",
+);
+
+@each $prop, $p in $props {
+	$sizes: ( 0, 1, 2, 3, 4, 5, 6, 7, 8 );
```
Comment (renatoagds) [inline projects/js-packages/components/components/text/style.module.scss:87]: This put a limit on us, that's the reason I vote for us to move for some `css-in-js` solution that helps with this integration between CSS and JS, but let's keep this for the future.

File: projects/js-packages/components/components/text/index.jsx (hunk 1 above)
Comment (retrofox) [reply projects/js-packages/components/components/text/index.jsx:59]: sorry I don't get it. `componentProps` is _the rest of the component props_, right? In the end, it pulls the props from the function argument.

File: projects/js-packages/components/components/text/style.module.scss (hunk 3 above)
Comment (retrofox) [reply projects/js-packages/components/components/text/style.module.scss:87]: agree, the style and js are not connect which means we cannot share data between them.

File: projects/js-packages/components/components/text/index.jsx (hunk 1 above)
Comment (renatoagds) [reply projects/js-packages/components/components/text/index.jsx:59]: @retrofox The rest props are designed to pass down props when you're using the `component` prop. The `component` props are designed to be used for the cases that you want the `text` styles and not use its default `component` for that moment.

Example of this usage:

https://github.com/Automattic/jetpack/blob/master/projects/packages/my-jetpack/_inc/components/product-detail-card/index.jsx#L218-L226

In this case, we're using the `Text` component only to get the `typography` styles, but the final component that renders is `ProductDetailButton`, and we pass some props that are for `ProductDetailButton` and not `Text`.

That's the reason we get what we need at `Text` and pass down only the rest.

The other reason is that `Unknown Prop Warning`: https://reactjs.org/warnings/unknown-prop.html 

Is a warning related to passing non-standard props for `DOM` elements, and we're mostly based on `DOM` elements.

File: projects/js-packages/components/components/text/style.module.scss (hunk 3 above)
Comment (renatoagds) [reply projects/js-packages/components/components/text/style.module.scss:87]: yes, at my P2: p9dueE-3Tq-p2 I highlighted this but waited to get more use cases to make this movement clearer.

File: projects/js-packages/components/components/text/index.jsx (hunk 1 above)
Comment (retrofox) [reply projects/js-packages/components/components/text/index.jsx:59]: I think I get that. The idea is picking those props that in some way belong to the `<text />` component. It means typography but also now spacing props. The rest of props should be pass down to the component:

```jsx
<Text
	component={ ProductDetailButton }
	onClick={ clickHandler }
	isLoading={ isFetching }
	isPrimary={ ! isBundle }
	href={ onClick ? undefined : addProductUrl }
	className={ styles[ 'checkout-button' ] }
	variant="body"
>
	Click me!
</Text>
``` 

Clearly, above, the props that belong to the Text component are `variant` and `component`, correct? We could do something like...

```jsx
<Text
	component={ ProductDetailButton }
	onClick={ clickHandler }
	href={ onClick ? undefined : addProductUrl }
	className={ styles[ 'checkout-button' ] }
	variant="body"
	mx={ 3 }
	pb={ 3 }
>
	Click me!
</Text>
```

Here, `component`, `variant`, `mx` and `pb` belong to Text component, and won't be propagated to the component instance.

Comment (retrofox) [reply projects/js-packages/components/components/text/index.jsx:59]: I'm deleting the spacing props in the line below:
```jsx
delete componentProps[ value ];
```

Comment (renatoagds) [reply projects/js-packages/components/components/text/index.jsx:59]: Oh! Missed the `delete` part, sorry. 🚀

Comment (retrofox) [reply projects/js-packages/components/components/text/index.jsx:59]: yeah, it deserves a comment tbh :yesyes:

File: projects/js-packages/components/components/text/stories/index.jsx (hunk 2 above)
Comment (retrofox) [reply projects/js-packages/components/components/text/stories/index.jsx:25]: https://github.com/Automattic/jetpack/pull/23682/commits/a57ef95b3bdf5e4c6042d236123493a79dca0078

File: projects/js-packages/components/components/text/index.jsx (hunk 1 above)
Comment (retrofox) [reply projects/js-packages/components/components/text/index.jsx:60]: https://github.com/Automattic/jetpack/pull/23682/commits/9f8b111cb0c488e40c33e15053a583e7d9cc7fd4

**Review by renatoagds — DISMISSED** [review]
LGTM

**Review by renatoagds — APPROVED** [review]

**Comment by keoshi** [comment]
Cool stuff!

Does `var( --spacing-base ) === 8px`?

**Comment by retrofox** [comment]
> Cool stuff!
> 
> Does `var( --spacing-base ) === 8px`?

Exactly. All text components are composed via `<Text />` component, and spacing wise, the unit is defined by `--spacing-base` which is 8px.

For instance, for Plans section and Status connection, the title is rendered by the `<H3 />` component:
markup | box-model
------|------
 | image

`<H3 />` simply creates a `<Text />` component that defines the typography and spacing:

```jsx
function H3( { children }) {
  return (
    <Text variant="headline-small" mb={ 3 }>
      { children }
    </Text>
  );
}
````

`mb={ 3 }` sets `margin-bottom: 24px` => `3 x 8px`

**Comment by retrofox** [comment]
> Great news! One last step: head over to your WordPress.com diff, D77784-code, and deploy it. Once you've done so, come back to this PR and add a comment with your changeset ID.

r242774-wpcom

**Pull request merged**